python/cudf_polars/cudf_polars/experimental/parallel.py

```diff
@@ -99,65 +86,69 @@ def _(ir: IR, rec: LowerIRTransformer) -> tuple[IR, MutableMapping[IR, Partition
 
 
-def lower_ir_graph(ir: IR) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+def task_graph(
+    ir: IR, partition_info: MutableMapping[IR, PartitionInfo]
+) -> tuple[MutableMapping[Any, Any], str | tuple[str, int]]:
     """
-    Rewrite an IR graph and extract partitioning information.
+    Construct a task graph for evaluation of an IR graph.
 
     Parameters
     ----------
     ir
         Root of the graph to rewrite.
+    partition_info
+        A mapping from all unique IR nodes to the
+        associated partitioning information.
 
     Returns
     -------
-    new_ir, partition_info
-        The rewritten graph, and a mapping from unique nodes
-        in the new graph to associated partitioning information.
+    graph
+        A Dask-compatible task graph for the entire
+        IR graph with root `ir`.
 
     Notes
     -----
-    This function traverses the unique nodes of the graph with
-    root `ir`, and applies :func:`lower_ir_node` to each node.
+    This function traverses the unique nodes of the
+    graph with root `ir`, and extracts the tasks for
+    each node with :func:`generate_ir_tasks`.
 
     See Also
     --------
-    lower_ir_node
+    generate_ir_tasks
     """
-    from cudf_polars.dsl.traversal import CachingVisitor
+    graph = reduce(
+        operator.or_,
+        (generate_ir_tasks(node, partition_info) for node in traversal(ir)),
+    )
 
-    mapper = CachingVisitor(lower_ir_node)
-    return mapper(ir)
+    key_name = get_key_name(ir)
+    partition_count = partition_info[ir].count
+    if partition_count > 1:
+        graph[key_name] = (_concat, list(partition_info[ir].keys(ir)))
+        return graph, key_name
+    else:
+        return graph, (key_name, 0)
 
 
-@singledispatch
-def generate_ir_tasks(
-    ir: IR, partition_info: MutableMapping[IR, PartitionInfo]
-) -> MutableMapping[Any, Any]:
-    """
-    Generate a task graph for evaluation of an IR node.
+def evaluate_dask(ir: IR) -> DataFrame:
+    """Evaluate an IR graph with Dask."""
+    from dask import get
 
-    Parameters
-    ----------
-    ir
-        IR node to generate tasks for.
-    partition_info
-        Partitioning information, obtained from :func:`lower_ir_graph`.
+    ir, partition_info = lower_ir_graph(ir)
 
-    Returns
-    -------
-    mapping
-        A (partial) dask task graph for the evaluation of an ir node.
+    graph, key = task_graph(ir, partition_info)
+    return get(graph, key)
 
-    Notes
-    -----
-    Task generation should only produce the tasks for the current node,
-    referring to child tasks by name.
 
-    See Also
-    --------
-    task_graph
-    """
-    raise AssertionError(f"Unhandled type {type(ir)}")  # pragma: no cover
+##
+## IR
+##
+
+
+@lower_ir_node.register(IR)
+def _(ir: IR, rec: LowerIRTransformer) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    # Single-partition default (see: _lower_ir_single)
+    return rec.state["default_mapper"](ir)
```

Comment:
Ah - I just realized I misunderstood [your earlier suggestion](https://github.com/rapidsai/cudf/pull/17262#discussion_r1856830391) and messed this up a bit.

>The idea is that we want a single-partition fallback for nodes where we're already defining a multi-partition handler.

Sort of. I just want a clean/intuitive way to fall back to "common" logic for any IR type. When we **don't** have a multi-partition handler defined for the IR type in question, I'd like to fall back to single-partition logic that is defined in one place. That logic would raise an error if there is **not** actually one partition. If we **do** have a multi-partition handler defined, it may still make sense for that handler to call that same single-partition logic in some cases (e.g. when support for specific options is missing, or there is only one partition).

A similar pattern will emerge for "partition-wise" operations. We are not going to want to repeat this logic all over the place, instead, we are going to want to call a `_lower_ir_partitionwise` function from multi-partition handler for the the IR type in question (e.g. Select).

Reply:
Note that I rolled back the "default_mapper" change for now (in favor of more-explicit handling for the fall-back case). Perhaps we can iron out our answer to this question in a PR that focuses on a "tricky" case like `Select`.